ops/testing.py

```diff
@@ -3249,19 +3271,80 @@ def send_signal(self, sig: Union[int, str], service_names: Iterable[str]):
             # conform with the real pebble api
             first_service = next(iter(service_names))
             message = f'cannot send signal to "{first_service}": invalid signal name "{sig}"'
-            body = {'type': 'error', 'status-code': 500, 'status': 'Internal Server Error',
-                    'result': {'message': message}}
-            raise pebble.APIError(
-                body=body,
-                code=500,
-                status='Internal Server Error',
-                message=message) from None
+            raise self._api_error(500, message)
 
     def get_checks(self, level=None, names=None):  # type:ignore
         raise NotImplementedError(self.get_checks)  # type:ignore
 
+    def notify(self, type: pebble.NoticeType, key: str, *,
+               data: Optional[Dict[str, str]] = None,
+               repeat_after: Optional[datetime.timedelta] = None) -> str:
+        notice_id, _ = self._notify(type, key, data=data, repeat_after=repeat_after)
+        return notice_id
+
+    def _notify(self, type: pebble.NoticeType, key: str, *,
+                data: Optional[Dict[str, str]] = None,
+                repeat_after: Optional[datetime.timedelta] = None) -> Tuple[str, bool]:
+        """Record an occurrence of a notice with the specified details.
+
+        Return a tuple of (notice_id, new_or_repeated).
+        """
+        if type != pebble.NoticeType.CUSTOM:
+            message = f'invalid type "{type.value}" (can only add "custom" notices)'
+            raise self._api_error(400, message)
+
+        # The shape of the code below is taken from State.AddNotice in Pebble.
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        uid = 0  # Hard-code UID as root (Pebble and charms run as root for now).
+
+        new_or_repeated = False
+        unique_key = (uid, type.value, key)
+        notice = self._notices.get(unique_key)
+        if notice is None:
+            # First occurrence of this notice uid+type+key
+            self._last_notice_id += 1
+            notice = pebble.Notice(
+                id=str(self._last_notice_id),
+                user_id=uid,
+                type=type,
+                key=key,
+                first_occurred=now,
+                last_occurred=now,
+                last_repeated=now,
+                expire_after=datetime.timedelta(days=7),
+                occurrences=1,
+            )
+            self._notices[unique_key] = notice
+            new_or_repeated = True
+        else:
+            # Additional occurrence, update existing notice
+            notice.occurrences += 1
```

Comment:
i suppose these might be why you didnt want to freeze notices.
and certainly the test side gets harder, but we did similar work elsewhere to, eg make Unit unmodifyable and Harness does the work to wrap that.
You certainly can use a frozen, use temp variables here and populate a new frozen notice in your dict when you are done.
Or use a non frozen notice-like, that gets cast to frozen so that *charm* code assumes they are frozen.
[knowing that it is safe to use this blob of data because nobody can modify it in your cache/behind your back is a useful property]
and we shouldnt be returning non frozen types from our dict without copying them anyway, because otherwise the charm under test might modify it, and then our dict was altered.

Reply:
Yep, see above -- I've changed this to reinstate the frozen=True, and I'm using `dataclasses.replace()` here to build a new notice.